### ISSUE TITLE
typescript definitions meta-issue

### ISSUE THREAD

**Issue by tamird** [dcodeIO/protobuf.js#718]
protobuf.js version: 6.6.5

First, thank you for including typescript support! I was delighted to discover that I could stop using https://github.com/sintef-9012/Proto2TypeScript.

However, there are a few problems with the generated and bundled TS definitions:
- [x] The included [`interface Long`](https://github.com/dcodeIO/protobuf.js/blob/6.6.5/index.d.ts#L2019) makes it impossible to chain `Long` methods on fields; e.g. `d.timestampNanos.toNumber()` becomes `d.timestampNanos as Long).toNumber()`
- [x] The [`generated type`](https://github.com/dcodeIO/protobuf.js/blob/6.6.5/cli/targets/static.js#L251:L256) of certain numeric fields is `number|Long`, which also makes it impossible to chain `Long` methods; i.e. it becomes necessary to coalesce or cast to `Long` before being able to use the field value.
- [x] The bundled type definitions include [hard-coded references](https://github.com/dcodeIO/protobuf.js/blob/6.6.5/index.d.ts#L1180:L1186) to [`Buffer`](https://nodejs.org/api/buffer.html), which is only available in Node.js. In [CockroachDB](https://github.com/cockroachdb/cockroach) we are using protobuf.js in the browser, and would like to avoid including type definitions for Node.js types which will not be available at run time.
- [x] The generated definitions should also include interfaces from which protobuf objects can be constructed (distinct from the protobuf objects themselves). This is partially related to #719.

**Comment by dcodeIO** [comment]
It seems that, no matter how this is done, that there'll always be some issues. For example, the minimal long interface has been introduced to handle the fact that long.js is optional. This solves the issue in most cases, but introduces others.

Similarly, previously all buffers were simply documented as being an `Uint8Array`, which is correct for newer versions of node, but requires explicit casts to use buffer functionality. So this has been changed to `Buffer` to make it work properly under node, and the typing referenced. However, `protobuf.BufferReader` and `protobuf.BufferWriter` are only used under node, though always included.

Related: https://github.com/dcodeIO/protobuf.js/issues/711

Long story short: I haven't found a way yet to make this work for all edge cases. Even generating multiple definition files, one for node and one for the browser, would introduce an inconvenient setup step to configure the correct file(s).

**Comment by tamird** [comment]
> te minimal long interface has been introduced to handle the fact that long.js is optional

What's the issue? Perhaps a reasonable solution is simply not to make it optional? I'm sure there is context I'm missing.

> Even generating multiple definition files, one for node and one for the browser, would introduce an inconvenient setup step to configure the correct file(s).

Seems to me that this is exactly what's needed, with a flag to `pbts` to import the desired one from generated code. You already do something similar with `{light,minimal}.{js,d.ts}`.

The other two issues I mentioned are with generated code - can they also be solved with flags to `pbts`?

**Comment by dcodeIO** [comment]
> Perhaps a reasonable solution is simply not to make it optional?

Well, not everyone needs full 64 bit support (JS numbers are safe up to 53 bits). If long.js isn't available, protobuf.js falls back to using numbers - and the user can safe quite a bit of bundle size.

**Comment by tamird** [comment]
Ah, fair enough. In that case, this can also be solved with another variant a la `{light,minimal}`, correct?

**Comment by dcodeIO** [comment]
Hmm, well, yeah. I'd favor something else, but that might be the only working solution. We'd then have `index, index-browser, light, light-browser, minimal, minimal-browser` or something.

**Comment by tamird** [comment]
That's a large matrix, but I don't know of another way. You'll also need to split the definitions (they currently are shared) to support real-long vs stub-long.

**Comment by tamird** [comment]
I've added another bullet to the list:

> The generated definitions should also include interfaces from which protobuf objects can be constructed (distinct from the protobuf objects themselves). This is partially related to #719.

**Comment by dcodeIO** [comment]
If you have any exact requirements for `fromObject` etc. feel free to join the discussion in 719. Currently, only constructors and `.create` are covered.

**Comment by tamird** [comment]
Ah, looks like you just implemented that in https://github.com/dcodeIO/protobuf.js/issues/717. Awesome.

**Comment by dcodeIO** [comment]
This commit removes minimal type definitions from library definitions and moves them to stubs instead:

* [stub-long.d.ts](https://github.com/dcodeIO/protobuf.js/blob/master/stub-long.d.ts)
* [stub-node.d.ts](https://github.com/dcodeIO/protobuf.js/blob/master/stub-node.d.ts)

Removing @types/node and/or @types/long would normally result in a tsc error, which can now be circumvented by referencing the stubs instead:

```ts
/// <reference path="protobuf.js/stub-long.d.ts" />
/// <reference path="protobuf.js/stub-node.d.ts" />
```

With this in place we could now either create a matrix of entry points that'd only differ in their references or leave this up to the developer as a manual step.

**Comment by tamird** [comment]
Nice. I'll try this stuff out shortly. Just one issue remaining, AFAICT, and I've marked that in the issue  description.

**Comment by dcodeIO** [comment]
The 2nd commit should cover this. It now emits `Long` instead of `$protobuf.Long`.

**Comment by tamird** [comment]
Yeah there's still the issue of number|Long, though.

**Comment by dcodeIO** [comment]
Well, this is still correct in some way. Even if you are not using long.js at all, you can still call a method with a proper `LongStub` (the thing now aliased as `Long`) as a parameter. As a return type, it doesn't make sense, though.

Possible workaround: Alias `Long` as a simple number, completely ignoring anything Long-like.

**Comment by dcodeIO** [comment]
This should do. `Long` now resolves as `number`, guaranteeing a consistent return type where long.js is not used. Likewise, where long.js is used, return types (not parameter types) can now be simplified to `Long` instead of `number|Long`. Looking into this now.

**Comment by dcodeIO** [comment]
This changes the return values. Note that messages properties can still be ´number|Long` because both is valid/supported as long as numbers are safe integers (<= 2^53).

**Comment by dcodeIO** [comment]
This now basically does what you want, I believe:

**Without long.js**
* Default is to output `number|Long` which resolves to just `number` because stub-long.d.ts just aliases `type Long = number`
* With `--strict-long` it outputs `Long`, which is shorter but stills resolves to just `number` because of the above

**With long.js**
* Default is to output `number|Long`
* With `--strict-long` it outputs `Long` only

**Comment by tamird** [comment]
Thanks so much! Going to close this in favour of the more specific #717 and #723.
